### PR TITLE
beancount-language-server 1.3.4 (new formula)

### PR DESCRIPTION
**Pull request by srstevenson**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/b/beancount-language-server.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class BeancountLanguageServer < Formula
+  desc "Language server for beancount files"
+  homepage "https://github.com/polarmutex/beancount-language-server"
+  url "https://github.com/polarmutex/beancount-language-server/archive/refs/tags/v1.3.4.tar.gz"
+  sha256 "d56f8d2d30f471b5f5011ffff9b55da9fecaec13f34a5f8f2ec0d6aa28cd3ded"
+  license "MIT"
+  head "https://github.com/polarmutex/beancount-language-server.git", branch: "main"
+
+  depends_on "rust" => :build
+  depends_on "beancount"
```
Comment (ZhongRuoyu) [inline Formula/b/beancount-language-server.rb:10]: Installing `beancount` does not sound strictly necessary according to https://github.com/polarmutex/beancount-language-server#requirements, so I think we could defer to the user to install this.

File: Formula/b/beancount-language-server.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  def install
+    system "cargo", "install", *std_cargo_args(path: "crates/lsp")
+  end
+
+  test do
+    system "#{bin}/beancount-language-server", "--help"
```
Comment (ZhongRuoyu) [inline Formula/b/beancount-language-server.rb:17]: Could you try with a test that covers some of its functionalities? Checks on version or help text almost never fail, even for broken installations.

For language servers specifically, you may want to check out [`cmake-language-server.rb`](https://github.com/Homebrew/homebrew-core/blob/6840a9fd5d0c225ca46919a0c2350932e0eba081/Formula/c/cmake-language-server.rb#L47-L60) and [`lua-language-server.rb`](https://github.com/Homebrew/homebrew-core/blob/6840a9fd5d0c225ca46919a0c2350932e0eba081/Formula/l/lua-language-server.rb#L44-L55).

Comment (srstevenson) [reply Formula/b/beancount-language-server.rb]: Thanks for the pointers! I've updated this to test that the server can be initialised.

**Review by ZhongRuoyu — APPROVED** [review]
Thanks @srstevenson!

File: Formula/b/beancount-language-server.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+          "rootUri": null,
+          "capabilities": {}
+        }
+      }
+    JSON
+
+    Open3.popen3("#{bin}/beancount-language-server", "--stdio") do |stdin, stdout|
+      stdin.write "Content-Length: #{json.size}\r\n\r\n#{json}"
+      assert_match(/^Content-Length: \d+/i, stdout.readline)
+    end
```
Comment (SMillerDev) [inline Formula/b/beancount-language-server.rb:33]: Can't this use `pipe_output` for capturing the in and output?

Comment (srstevenson) [reply Formula/b/beancount-language-server.rb:33]: I tried using `pipe_output`, and the server panics on trying to write to a disconnected channel. I found the `open3` pattern in the formulae for other language servers, which I assume to be for the same reason based on explicit `sleep` calls before reading from stdout: https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-core+Open3.popen3+path%3A*-language-server.rb&type=code

**Review by chenrui333 — APPROVED** [review]

**Comment by chenrui333** [comment]
@srstevenson, thanks for your contribution to Homebrew! :tada: 🥇

Without awesome contributors like you, it would be impossible to maintain Homebrew to the high level of quality users have come to expect. Thank you!!!!

**Pull request merged**